### PR TITLE
Generalize `AnnTip`

### PR DESCRIPTION
**Pull request by edsko**
Previously `AnnTip` looked like

```haskell
 data AnnTip blk = AnnTip {
       annTipSlotNo  :: !SlotNo
     , annTipHash    :: !(HeaderHash blk)
     , annTipBlockNo :: !BlockNo
     , annTipInfo    :: !(TipInfo blk)
     }
```

That double use of `blk` is problematic. Consider what happens when we instantiate this to `HardForkBlock`: the hash of a hard fork block is a hash from one of the eras (Byron hash, Shelley hash, ..), and similarly the `TipInfo` of a `HardForkBlock` is the tip info from a particular era. But if we instantiate this twice, independently, then we lose the information that these two must be _related_; it doesns't make sense to have a block with a Byron header but with Shelley info.

Therefore in this PR we delete `annTipHash` altogether, and instead insist that we must be able to _derive_ the `HeaderHash `from the `TipInfo`.

The first 4 commits in this PR are all designed specifically to ensure that introducing this change does not break backwards compatibility. The encoders/decoders for `AnnTip` still encode the information in precisely the same order; only the in-memory representation is now different.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs [hunk 1]
```diff
@@ -442,53 +450,82 @@ validateHeader cfg ledgerView hdr st = do
...
+                    => (HeaderHash blk -> Encoding)
+                    -> (AnnTip     blk -> Encoding)
+defaultEncodeAnnTip encodeHash AnnTip{..} = mconcat [
       encodeListLen 4
     , encode     annTipSlotNo
-    , encodeHash annTipHash
+    , encodeHash annTipInfo
     , encode     annTipBlockNo
-    , encodeInfo annTipInfo
+      -- TODO: Useless field. Remove when OK to break binary compatibility.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs:461]: How will we think of this when the opportunity is there to break binary compatibility?

Do we actually expect this to happen?

Alternatively, we do the simple naive thing in the encoder and make the decoder backwards compatible by branching on the size (4 -> old, 3 -> new).

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs:461]: Once we move the hard fork combinator we break binary compatibility, I think that's the natural point to introduce this. When we create the tickets for the various TODOs in #2034 and create one to switch the main code base over to `HardForkBlock [Byron, Shelley]`, maybe let's change it then?

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs:461]: https://github.com/input-output-hk/ouroboros-network/issues/2062

**Comment by edsko** [comment]
bors merge

**Pull request merged**